### ISSUE TITLE
NullReferenceException - Regression from 1.4.0 to 1.5.0

### ISSUE THREAD

**Issue by bslbckr** [agileobjects/AgileMapper#167]
Hi,

we are currently trying to update our code-base from version 1.4.0 to 1.5.0 as we need some of the features only available in 1.5.0. Unfortunately, after upgrading the depdencies several of our unit-tests fail, due to different behavior between the two versions.  Even worse, we did not come up with a quick solution to the problems. 

I tried to create a [fiddle](https://dotnetfiddle.net/FcpWUO), to isolate and illustrate the problem but I receive a different exception there as in our code. In the fiddle I get a `NullReferenceException` whereas our code throws a `InvalidOperationException: No coercion operator is defined between types 'System.Int32' and '`

If you change the fiddle back to 1.4.0 three mapping plans will be generated.

In our code-base we get the following stack-trace:
```
 1)  ParameterServiceServerTest.CoreToViaConfigTest.TestMapListValues<Int32>(input: { "intVals": { "values": [ 1, 2, 3 ] } }, expected: [1, 2, 3])
      Duration: 502 ms

      Message: 
        System.InvalidOperationException : No coercion operator is defined between types 'System.Int32' and 'Google.Protobuf.Collections.RepeatedField`1[System.Int32]'.
      Stack Trace: 
        Expression.GetUserDefinedCoercionOrThrow(ExpressionType coercionType, Expression expression, Type convertToType)
        Expression.Convert(Expression expression, Type type, MethodInfo method)
        Member.GetAccess(Expression instance)
        Enumerable.Aggregate[TSource,TAccumulate](IEnumerable`1 source, TAccumulate seed, Func`3 func)
        ParametersSwapper.GetSimpleTypesMappingContextInfo(SwapArgs swapArgs)
        ParametersSwapper.ReplaceParameters(SwapArgs swapArgs, Func`2[] parameterFactories)
        ConfiguredLambdaInfo.GetBody(IMemberMapperData mapperData, Nullable`1 position, QualifiedMember targetMember)
        ConfiguredDataSourceFactory.Create(IMemberMapperData mapperData)
        PublicEnumerableExtensions.Project[TItem,TArg,TResult](IEnumerable`1 items, TArg argument, Func`3 projector)+MoveNext()
        LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
        EnumerableHelpers.ToArray[T](IEnumerable`1 source)
        Enumerable.ToArray[TSource](IEnumerable`1 source)
        DataSourceFindContext.get_ConfiguredDataSources()
        ConfiguredDataSourcesFactory.Create(DataSourceFindContext context)+MoveNext()
        DataSourceSetFactory.EnumerateDataSources(DataSourceFindContext context)+MoveNext()
        LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
        EnumerableHelpers.ToArray[T](IEnumerable`1 source)
        Enumerable.ToArray[TSource](IEnumerable`1 source)
        DataSourceSetFactory.CreateFor(DataSourceFindContext findContext)
        MemberPopulatorFactory.Create(MemberPopulationContext context)
        PublicEnumerableExtensions.Project[TItem,TArg,TResult](IEnumerable`1 items, TArg argument, Func`3 projector)+MoveNext()
        PublicEnumerableExtensions.Filter[TItem,TArg](IEnumerable`1 items, TArg argument, Func`3 predicate)+MoveNext()
        PopulationExpressionFactoryBase.GetPopulationsAndCallbacks(IObjectMappingData mappingData)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        Enumerable.ToList[TSource](IEnumerable`1 source)
        PopulationExpressionFactoryBase.GetPopulation(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
        ComplexTypeDataSource.Create(IObjectMappingData mappingData)
        DataSourceSetFactory.CreateFor(IObjectMappingData rootMappingData)
        ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.GetOrCreateMapper()
        MappingFactory.GetInlineMappingBlock(IObjectMappingData mappingData, MappingValues mappingValues, Expression createMappingDataCall)
        EnumerablePopulationBuilder.GetElementPopulation(IPopulationLoopData loopData, IObjectMappingData mappingData)
        PopulationLoopDataExtensions.BuildPopulationLoop[TLoopData](TLoopData loopData, EnumerablePopulationBuilder builder, IObjectMappingData mappingData, Func`3 elementPopulationFactory)
        EnumerablePopulationBuilder.BuildPopulationLoop(Func`3 elementPopulationFactory, IObjectMappingData mappingData)
        CopySourceEnumerablePopulationStrategy.Create(EnumerablePopulationBuilder builder, IObjectMappingData enumerableMappingData)
        EnumerableMappingExpressionFactory.GetObjectPopulation(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.GetConfiguredToTargetDataSourceMappings(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
        MappingRootDataSourceFactoryBase.CreateFor(IObjectMappingData mappingData)
        DataSourceSetFactory.CreateFor(IObjectMappingData rootMappingData)
        ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.GetOrCreateMapper()
        ArrayCache`2.GetOrAdd(TKey key, Func`2 valueFactory)
        ObjectMapperFactory.GetOrCreateRoot[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.ctor(TSource source, TTarget target, Nullable`1 enumerableIndex, MappingTypes mappingTypes, IMappingContext mappingContext, IObjectMappingData declaredTypeMappingData, IObjectMappingData parent, Boolean createMapper)
        ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, MappingTypes mappingTypes, IMappingContext mappingContext, Boolean createMapper)
        ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, IMappingContext mappingContext, Boolean createMapper)
        MappingExecutor`1.PerformMapping[TTarget](TTarget target)
        MappingExecutor`1.ToANew[TResult]()
        CoreToViaConfigTest.TestMapListValues[T](ValueWrapper input, IList`1 expected) line 138

    2)  ParameterServiceServerTest.CoreToViaConfigTest.TestMapListValues<Double>(input: { "doubleVals": { "values": [ 1, 2, 3 ] } }, expected: [1, 2, 3])
      Duration: 18 ms

      Message: 
        System.InvalidOperationException : No coercion operator is defined between types 'System.Int32' and 'Google.Protobuf.Collections.RepeatedField`1[System.Int32]'.
      Stack Trace: 
        Expression.GetUserDefinedCoercionOrThrow(ExpressionType coercionType, Expression expression, Type convertToType)
        Expression.Convert(Expression expression, Type type, MethodInfo method)
        Member.GetAccess(Expression instance)
        Enumerable.Aggregate[TSource,TAccumulate](IEnumerable`1 source, TAccumulate seed, Func`3 func)
        ParametersSwapper.GetSimpleTypesMappingContextInfo(SwapArgs swapArgs)
        ParametersSwapper.ReplaceParameters(SwapArgs swapArgs, Func`2[] parameterFactories)
        ConfiguredLambdaInfo.GetBody(IMemberMapperData mapperData, Nullable`1 position, QualifiedMember targetMember)
        ConfiguredDataSourceFactory.Create(IMemberMapperData mapperData)
        PublicEnumerableExtensions.Project[TItem,TArg,TResult](IEnumerable`1 items, TArg argument, Func`3 projector)+MoveNext()
        LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
        EnumerableHelpers.ToArray[T](IEnumerable`1 source)
        Enumerable.ToArray[TSource](IEnumerable`1 source)
        DataSourceFindContext.get_ConfiguredDataSources()
        ConfiguredDataSourcesFactory.Create(DataSourceFindContext context)+MoveNext()
        DataSourceSetFactory.EnumerateDataSources(DataSourceFindContext context)+MoveNext()
        LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
        EnumerableHelpers.ToArray[T](IEnumerable`1 source)
        Enumerable.ToArray[TSource](IEnumerable`1 source)
        DataSourceSetFactory.CreateFor(DataSourceFindContext findContext)
        MemberPopulatorFactory.Create(MemberPopulationContext context)
        PublicEnumerableExtensions.Project[TItem,TArg,TResult](IEnumerable`1 items, TArg argument, Func`3 projector)+MoveNext()
        PublicEnumerableExtensions.Filter[TItem,TArg](IEnumerable`1 items, TArg argument, Func`3 predicate)+MoveNext()
        PopulationExpressionFactoryBase.GetPopulationsAndCallbacks(IObjectMappingData mappingData)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        Enumerable.ToList[TSource](IEnumerable`1 source)
        PopulationExpressionFactoryBase.GetPopulation(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
        ComplexTypeDataSource.Create(IObjectMappingData mappingData)
        DataSourceSetFactory.CreateFor(IObjectMappingData rootMappingData)
        ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.GetOrCreateMapper()
        MappingFactory.GetInlineMappingBlock(IObjectMappingData mappingData, MappingValues mappingValues, Expression createMappingDataCall)
        EnumerablePopulationBuilder.GetElementPopulation(IPopulationLoopData loopData, IObjectMappingData mappingData)
        PopulationLoopDataExtensions.BuildPopulationLoop[TLoopData](TLoopData loopData, EnumerablePopulationBuilder builder, IObjectMappingData mappingData, Func`3 elementPopulationFactory)
        EnumerablePopulationBuilder.BuildPopulationLoop(Func`3 elementPopulationFactory, IObjectMappingData mappingData)
        CopySourceEnumerablePopulationStrategy.Create(EnumerablePopulationBuilder builder, IObjectMappingData enumerableMappingData)
        EnumerableMappingExpressionFactory.GetObjectPopulation(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.GetConfiguredToTargetDataSourceMappings(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
        MappingRootDataSourceFactoryBase.CreateFor(IObjectMappingData mappingData)
        DataSourceSetFactory.CreateFor(IObjectMappingData rootMappingData)
        ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.GetOrCreateMapper()
        ArrayCache`2.GetOrAdd(TKey key, Func`2 valueFactory)
        ObjectMapperFactory.GetOrCreateRoot[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.ctor(TSource source, TTarget target, Nullable`1 enumerableIndex, MappingTypes mappingTypes, IMappingContext mappingContext, IObjectMappingData declaredTypeMappingData, IObjectMappingData parent, Boolean createMapper)
        ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, MappingTypes mappingTypes, IMappingContext mappingContext, Boolean createMapper)
        ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, IMappingContext mappingContext, Boolean createMapper)
        MappingExecutor`1.PerformMapping[TTarget](TTarget target)
        MappingExecutor`1.ToANew[TResult]()
        CoreToViaConfigTest.TestMapListValues[T](ValueWrapper input, IList`1 expected) line 138

    3)  ParameterServiceServerTest.CoreToViaConfigTest.TestMapListValues<Boolean>(input: { "booleanVals": { "values": [ true, true, false, true ] } }, expected: [True, True, False, True])
      Duration: 19 ms

      Message: 
        System.InvalidOperationException : No coercion operator is defined between types 'System.Int32' and 'Google.Protobuf.Collections.RepeatedField`1[System.Int32]'.
      Stack Trace: 
        Expression.GetUserDefinedCoercionOrThrow(ExpressionType coercionType, Expression expression, Type convertToType)
        Expression.Convert(Expression expression, Type type, MethodInfo method)
        Member.GetAccess(Expression instance)
        Enumerable.Aggregate[TSource,TAccumulate](IEnumerable`1 source, TAccumulate seed, Func`3 func)
        ParametersSwapper.GetSimpleTypesMappingContextInfo(SwapArgs swapArgs)
        ParametersSwapper.ReplaceParameters(SwapArgs swapArgs, Func`2[] parameterFactories)
        ConfiguredLambdaInfo.GetBody(IMemberMapperData mapperData, Nullable`1 position, QualifiedMember targetMember)
        ConfiguredDataSourceFactory.Create(IMemberMapperData mapperData)
        PublicEnumerableExtensions.Project[TItem,TArg,TResult](IEnumerable`1 items, TArg argument, Func`3 projector)+MoveNext()
        LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
        EnumerableHelpers.ToArray[T](IEnumerable`1 source)
        Enumerable.ToArray[TSource](IEnumerable`1 source)
        DataSourceFindContext.get_ConfiguredDataSources()
        ConfiguredDataSourcesFactory.Create(DataSourceFindContext context)+MoveNext()
        DataSourceSetFactory.EnumerateDataSources(DataSourceFindContext context)+MoveNext()
        LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
        EnumerableHelpers.ToArray[T](IEnumerable`1 source)
        Enumerable.ToArray[TSource](IEnumerable`1 source)
        DataSourceSetFactory.CreateFor(DataSourceFindContext findContext)
        MemberPopulatorFactory.Create(MemberPopulationContext context)
        PublicEnumerableExtensions.Project[TItem,TArg,TResult](IEnumerable`1 items, TArg argument, Func`3 projector)+MoveNext()
        PublicEnumerableExtensions.Filter[TItem,TArg](IEnumerable`1 items, TArg argument, Func`3 predicate)+MoveNext()
        PopulationExpressionFactoryBase.GetPopulationsAndCallbacks(IObjectMappingData mappingData)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        Enumerable.ToList[TSource](IEnumerable`1 source)
        PopulationExpressionFactoryBase.GetPopulation(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
        ComplexTypeDataSource.Create(IObjectMappingData mappingData)
        DataSourceSetFactory.CreateFor(IObjectMappingData rootMappingData)
        ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.GetOrCreateMapper()
        MappingFactory.GetInlineMappingBlock(IObjectMappingData mappingData, MappingValues mappingValues, Expression createMappingDataCall)
        EnumerablePopulationBuilder.GetElementPopulation(IPopulationLoopData loopData, IObjectMappingData mappingData)
        PopulationLoopDataExtensions.BuildPopulationLoop[TLoopData](TLoopData loopData, EnumerablePopulationBuilder builder, IObjectMappingData mappingData, Func`3 elementPopulationFactory)
        EnumerablePopulationBuilder.BuildPopulationLoop(Func`3 elementPopulationFactory, IObjectMappingData mappingData)
        CopySourceEnumerablePopulationStrategy.Create(EnumerablePopulationBuilder builder, IObjectMappingData enumerableMappingData)
        EnumerableMappingExpressionFactory.GetObjectPopulation(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.GetConfiguredToTargetDataSourceMappings(MappingCreationContext context)+MoveNext()
        List`1.AddEnumerable(IEnumerable`1 enumerable)
        List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
        MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
        MappingRootDataSourceFactoryBase.CreateFor(IObjectMappingData mappingData)
        DataSourceSetFactory.CreateFor(IObjectMappingData rootMappingData)
        ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.GetOrCreateMapper()
        ArrayCache`2.GetOrAdd(TKey key, Func`2 valueFactory)
        ObjectMapperFactory.GetOrCreateRoot[TSource,TTarget](ObjectMappingData`2 mappingData)
        ObjectMappingData`2.ctor(TSource source, TTarget target, Nullable`1 enumerableIndex, MappingTypes mappingTypes, IMappingContext mappingContext, IObjectMappingData declaredTypeMappingData, IObjectMappingData parent, Boolean createMapper)
        ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, MappingTypes mappingTypes, IMappingContext mappingContext, Boolean createMapper)
        ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, IMappingContext mappingContext, Boolean createMapper)
        MappingExecutor`1.PerformMapping[TTarget](TTarget target)
        MappingExecutor`1.ToANew[TResult]()
        CoreToViaConfigTest.TestMapListValues[T](ValueWrapper input, IList`1 expected) line 138
```

**Comment by SteveWilkes** [comment]
Hello!

I can't reproduce either of the exceptions from your two fiddles in the latest v1.6 branch code, so hopefully that means other fixes and changes have addressed those errors. We'll see for sure when the next preview comes out :)

Cheers,

Steve

**Comment by bslbckr** [comment]
Even better if this is already fixed. 
Unfortunately, I don't remember which versions I tried with the fiddle. Nevertheless I think that this scenario will make a good candidate for a regression test.

**Comment by bslbckr** [comment]
would you mind publishing an updated preview-version of the mapper such that we can try it out?

**Comment by SteveWilkes** [comment]
This is fixed as of v1.6-preview4, which is now [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper). Thanks again!

**Comment by SteveWilkes** [comment]
The fix for this is included in v1.6, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper/1.6.0).

Thanks again for the feedback!
